Implement name lookups by hash

domob1812 opened the pull request:
This set of changes implements https://github.com/namecoin/namecoin-core/issues/329:  A new option `-namehashindex` allows running Namecoin Core in a way that keeps an additional database of name hashes to preimages for all registered names.  When this is enabled, the new `byHash` option can be passed to `name_show` and `name_history` to request name lookup by hash rather than direct.

This is useful because it allows using an external node to check if a given name is registered or not, without revealing the name itself if it does not exist yet.  This may be useful for Electrum (and perhaps other light clients) in the future.

domob1812 commented:
@JeremyRand, please take a look and let me know if that is what you need for Electrum.  Thanks!

domob1812 commented:
@JeremyRand, please take a look at this when you have time.

domob1812 commented:
@JeremyRand Any news on this?  Do you not want to use that for Electrum anymore?

JeremyRand commented:
> @JeremyRand Any news on this? Do you not want to use that for Electrum anymore?

@domob1812 We're still going to use it; in fact it's a prerequisite to one of the NLnet subtasks we just filed for.  So I'll test it out prior to that NLnet subtask's work beginning; hopefully I can get to it within the next few weeks, just been busy.

domob1812 commented:
Ok cool, thanks!  I'd just like to merge this rather sooner than later, to avoid having to rebase this repeatedly for upstream changes.  It has been open for quite some time already.

domob1812 commented:
@JeremyRand Friendly ping.  Even if you do not yet work immediately on whatever tasks need this, could you perhaps just try it out by itself and let me know if it can be merged as is?  We can always tweak and fix it later anyway in case something is not exactly as needed, as long as you can confirm the general direction is what you need.

JeremyRand commented:
> @JeremyRand Friendly ping. Even if you do not yet work immediately on whatever tasks need this, could you perhaps just try it out by itself and let me know if it can be merged as is? We can always tweak and fix it later anyway in case something is not exactly as needed, as long as you can confirm the general direction is what you need.

@domob1812 I'll test it today (starting a build now).

JeremyRand commented on src/index/namehash.h:
```diff
@@ -0,0 +1,73 @@
+static constexpr int64_t MAX_NAMEHASH_CACHE = 1024;
+
+/**
+ * This keeps an index of SHA-256-hashes of names to the corresponding preimage
```
If I'm reading the code correctly, `SHA-256` should be `SHA-256d` in this comment.

domob1812 replied:
Yes indeed, thanks for pointing this out!

JeremyRand commented:
"Namecoin release nodes" (in commit message) is a typo, should be "Namecoin release notes"

JeremyRand commented:
ACK ff174027002d0bc1497c636657052a436b9c5579, subject to the `SHA-256` and `nodes` typos being fixed.  Everything appears to work fine in my testing.

A couple of non-blocking inquiries:

* I notice that entering a hash as hex only works when the name encoding is set to hex, meaning that the result JSON object will encode the preimage as hex.  Is this intentional?  I don't have any objection to this (if anything I think it's good to nudge users toward hex encoding, so I'd favor keeping it like this), just curious if this was intended.
* It might be nice to have the `name_show` result JSON object include a field containing the SHA-256d hash of the name, which would make it more convenient for users looking at a result object to know by eye which hash can be used to look up the name.  (That doesn't have to be in this PR, but might be worth considering in a separate issue.)

JeremyRand commented:
(Once you've fixed those 2 typos and fixed the merge conflict, feel free to merge without asking me to review again.)

domob1812 commented:
Thanks for the feedback, I've fixed the two typos and rebased onto current `master`.

Since the hash is passed in the `name` argument, the code uses (intentionally) the same logic for decoding it as it would for normal names.  Hence you can in theory pass a hash also e.g. with ASCII-encoding, but only if it would consist only of ASCII bytes.  (Which means that in practice you are forced to set the encoding to `hex` for using this.)  I think removing this and always interpreting the `name` as hex when looking up by hash would introduce confusing "special rules" we do not want, so I'll keep it like that.

Regarding returning the hash:  I believe this is likely something that "looks" like nice-to-have, but hasn't really any practical relevance.  Or do you have an actual application in mind where this would be useful?  Mainly I think that if you already have the `name_show` result from a lookup, you will never be in need to get the hash anymore, would you?  Thus I'll defer this for later.

domob1812 commented:
Oh, and since 0.20 was branched off while this was open, I've changed the changelog entry for this to 0.21.
